cl/recap/tasks.py

```diff
@@ -1512,6 +1509,46 @@ def send_docket_to_webhook(d_pk: int, since: datetime, epq_pk: int) -> None:
             webhook.save()
 
 
+def first_time_users(
+    docket_pk: int,
+    email_recipients: list[dict[str, str | list[str]]],
+) -> list[int]:
+    """Get the recap.email recipients and determine if it's the first time that
+     a recap.email notification comes in for a case-user.
+
+    :param email_recipients: List of dicts that contains the notification
+    email recipients in the format: "name": name, "email_addresses": [""]
+    :param docket_pk: The PK of the docket related to the notification
+    :return: List of recap.email user_pks if it's the first time the
+     recap.email notification comes in for this case-user
+    """
+    # Extract all email addresses from email_recipients
+    email_addresses = [
+        email
+        for email_address in email_recipients
+        for email in email_address["email_addresses"]
+    ]
+    # Select only @recap.email addresses
+    recap_email_recipients = [
+        recap_email
+        for recap_email in email_addresses
+        if "@recap.email" in recap_email
+    ]
+
+    recap_users = User.objects.filter(
+        profile__recap_email__in=recap_email_recipients
+    )
+    first_time_recipients = []
+    for user in recap_users:
+        docket_alert = DocketAlert.objects.filter(
+            docket_id=docket_pk, user_id=user.pk
+        )
+        if not docket_alert.exists():
+            first_time_recipients.append(user.pk)
```

Comment:
I'm sorry to ask for more revisions, but this is a performance-critical part of the code since we send so many messages. I think there's one more tweak that could help here. Instead of sending `first_time_recipients` as the user_pks, if we send it as the email addresses, it would allow us to avoid more queries in alerts/tasks.py. This:

```
user = User.objects.get(email=email_address)
if user.pk in first_time_recipients:
```

Could become:
```
if email_address in first_time_recipients:
```

I *think* that will help further simplify this and will remove a few more queries. I keep looking carefully at this because we're doing so many lookups, it makes me think something is wrong. Parse the email, then look up the user pks, we then use those to look up the email address, and then we use them again to look up the auto-subscribe info. It's a look of queries against the user table, which makes me think something is off. 

The above should simplify though a bit. The other thing to do would be to pass more info through to our `send_docket_alerts` task, so that it has all the context it needs and doesn't have to look it up again. 

See what you think. It feels messy and round-about to me, but I think it's close to it's simplest form.

Reply:
Thank you! yes, this simplified things and reduced queries,  I renamed `first_time_users` to `get_recap_email_recipients` because now we pass all the recap.email addresses to `send_docket_alert`, so checking if it's a first_email_recipient for that case is performed now in `make_alert_messages`, if exists a DocketAlert for `recap_email_recipients` we avoid sending the first email, so now we only make a query for `recap_email_recipients` to `UserProfile` to look up the auto-subscribe and Docket Alert and one for not `first-time subscriptors` to get the `User` based on the email address and get the DocketAlert to pass to the email context.